mutable_set/mutable_set.mbt

```diff
@@ -0,0 +1,568 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// MutableSet is an open-addressing hashset based on robin hood hash.
+pub struct MutableSet[T] {
+  priv mut directory : Array[Option[Item[T]]]
+  priv mut real_size : Int
+  priv mut effective_size : Int
+}
+
+priv struct Item[T] {
+  mut key : T
+  mut probe_length : Int
+} derive(Debug, Show)
+
+/// Initialize MutableSet by specifying initial capacity.
+pub fn MutableSet::with_capacity[T](initial_capacity : Int) -> MutableSet[T] {
+  {
+    directory: if initial_capacity <= 0 || initial_capacity > 2147483647 {
+      @array.new(8, fn() -> Option[Item[T]] { None })
+    } else {
+      @array.new(initial_capacity, fn() -> Option[Item[T]] { None })
+    },
+    real_size: 0,
+    effective_size: 0,
+  }
+}
+
+/// Initialize MutableSet with default capacity 3.
+pub fn MutableSet::new[T]() -> MutableSet[T] {
+  MutableSet::with_capacity(8)
+}
+
+/// Initialize a MutableSet from array.
+pub fn MutableSet::from_array[T : Hash + Eq](array : Array[T]) -> MutableSet[T] {
+  let set = MutableSet::with_capacity(array.length())
+  array.iter(fn(key : T) -> Unit { set.insert(key) })
+  set
+}
+
+/// Convert MutableSet to Array.
+/// 
+/// # Example
+/// 
+/// ```
+/// println(MutableSet::[1, 1, 2, 2, 3, 3].to_array())
+/// // output: [3, 1, 2]
+/// ```
+pub fn to_array[T : Default](self : MutableSet[T]) -> Array[T] {
+  // If Array has a filter method, we can implement it like this:
+  // let array : Array[Option[T]] = @array.new(self.size(), fn () -> Option[T] { None })
+  // let mut index = 0
+  // self.iter(
+  //   fn(key : T) -> Unit {
+  //     array[index] = Some(key)
+  //     index = index + 1
+  //   },
+  // )
+  // array.filter(fn(item : Option[T]) -> Bool { item.is_empty() })
+  self.to_list().to_array()
+}
+
+/// Initialize a MutableSet from list.
+pub fn MutableSet::from_list[T : Hash + Eq](list : List[T]) -> MutableSet[T] {
+  let set = MutableSet::with_capacity(list.length())
+  list.iter(fn(key : T) -> Unit { set.insert(key) })
+  set
+}
+
+/// Convert MutableSet to List.
+/// 
+/// # Example
+/// 
+/// ```
+/// println(MutableSet::[1, 1, 2, 2, 3, 3].to_list().sort())
+/// // output: Cons(1, Cons(2, Cons(3, Nil)))
+/// ```
+pub fn to_list[T](self : MutableSet[T]) -> List[T] {
+  self.fold(
+    fn(key : T, list : List[T]) -> List[T] { Cons(key, list) },
+    ~init=Nil,
+  )
+}
+
+/// Remove all elements from MutableSet.
+/// 
+/// # Example
+/// 
+/// ```
+/// let set = MutableSet::[1, 1, 2, 2, 3, 3]
+/// set.clear()
+/// println(set.to_list())
+/// // output: Nil
+/// ```
+pub fn clear[T](self : MutableSet[T]) -> Unit {
+  let mut index = 0
+  while index < self.directory.length() {
+    self.directory[index] = None
+    index = index + 1
+  }
+  self.effective_size = 0
+  self.real_size = 0
+}
+
+/// Returns a copy of MutableSet.
+pub fn copy[T](self : MutableSet[T]) -> MutableSet[T] {
+  {
+    directory: self.directory,
```

Comment:
```moonbit
test "union" {
  let a = MutableSet::[3, 4, 5]
  let b = MutableSet::[3, 4, 5]
  let c = a.union(b)
  a.clear()
  @assertion.assert_true(c.exist(3))?
}
```

Reply:
I recklessly thought that MoonBit would copy direcory here. After testing, the basic type will perform copying, so I need to copy the directory manually.

Reply:
We may need `Array::copy` @Yu-zh